### ISSUE TITLE
Loss is nan during training with OpenCL runtime but not with CPU=1

### ISSUE THREAD

**Issue by russnes** [tinygrad/tinygrad#9608]
When training, loss is `nan` randomly throughout epochs if I use the GPU backend, while it works with `CPU=1`. Interestingly, if I run the GPU backend with `DEBUG=2`, then the loss stays not-nan.

GPU: AMD Radeon RX 460
OS: Linux
tinygrad version: current master
Python: 3.12

**With GPU:**

    $ python minimal_tinygrad_opencl.py 
    Current device: GPU
    Epoch 1: Loss: 0.9986
    Epoch 2: Loss: nan
    Epoch 3: Loss: nan
    Epoch 4: Loss: nan
    Epoch 5: Loss: nan
    Epoch 6: Loss: nan
    Epoch 7: Loss: 0.4104
    Epoch 8: Loss: 0.3511
    Epoch 9: Loss: 0.2487
    Epoch 10: Loss: nan
    Epoch 11: Loss: 0.1551
    Epoch 12: Loss: nan
    Epoch 13: Loss: nan
    Epoch 14: Loss: nan
    Epoch 15: Loss: 0.0267
    Epoch 16: Loss: nan
    Epoch 17: Loss: nan
    Epoch 18: Loss: nan
    Epoch 19: Loss: 0.0113
    Epoch 20: Loss: nan
    Epoch 21: Loss: nan
    Epoch 22: Loss: 0.0032
    Epoch 23: Loss: 0.0021
    Epoch 24: Loss: nan
    Epoch 25: Loss: nan
    Epoch 26: Loss: nan
    Epoch 27: Loss: nan
    Epoch 28: Loss: nan
    Epoch 29: Loss: 0.0011
    Epoch 30: Loss: 0.0011
    Epoch 31: Loss: nan
    Epoch 32: Loss: nan
    Epoch 33: Loss: 0.0008
    Epoch 34: Loss: nan
    Epoch 35: Loss: nan
    Epoch 36: Loss: nan
    Epoch 37: Loss: nan
    Epoch 38: Loss: nan
    Epoch 39: Loss: nan
    Epoch 40: Loss: nan
    Epoch 41: Loss: nan
    Epoch 42: Loss: nan
    Epoch 43: Loss: nan
    Epoch 44: Loss: nan
    Epoch 45: Loss: nan
    Epoch 46: Loss: 0.0001
    Epoch 47: Loss: nan
    Epoch 48: Loss: 0.0001
    Epoch 49: Loss: nan
    Epoch 50: Loss: nan

**With CPU:**

    $ CPU=1 python minimal_tinygrad_opencl.py 
    Current device: CPU
    Epoch 1: Loss: 0.7119
    Epoch 2: Loss: 1.0425
    Epoch 3: Loss: 0.4563
    Epoch 4: Loss: 0.5740
    Epoch 5: Loss: 0.4090
    Epoch 6: Loss: 0.2131
    Epoch 7: Loss: 0.1569
    Epoch 8: Loss: 0.1478
    Epoch 9: Loss: 0.1142
    Epoch 10: Loss: 0.0621
    Epoch 11: Loss: 0.0307
    Epoch 12: Loss: 0.0207
    Epoch 13: Loss: 0.0165
    Epoch 14: Loss: 0.0146
    Epoch 15: Loss: 0.0131
    Epoch 16: Loss: 0.0114
    Epoch 17: Loss: 0.0093
    Epoch 18: Loss: 0.0070
    Epoch 19: Loss: 0.0050
    Epoch 20: Loss: 0.0035
    Epoch 21: Loss: 0.0024
    Epoch 22: Loss: 0.0016
    Epoch 23: Loss: 0.0012
    Epoch 24: Loss: 0.0009
    Epoch 25: Loss: 0.0007
    Epoch 26: Loss: 0.0006
    Epoch 27: Loss: 0.0005
    Epoch 28: Loss: 0.0005
    Epoch 29: Loss: 0.0005
    Epoch 30: Loss: 0.0005
    Epoch 31: Loss: 0.0005
    Epoch 32: Loss: 0.0005
    Epoch 33: Loss: 0.0005
    Epoch 34: Loss: 0.0005
    Epoch 35: Loss: 0.0005
    Epoch 36: Loss: 0.0004
    Epoch 37: Loss: 0.0004
    Epoch 38: Loss: 0.0004
    Epoch 39: Loss: 0.0004
    Epoch 40: Loss: 0.0004
    Epoch 41: Loss: 0.0003
    Epoch 42: Loss: 0.0003
    Epoch 43: Loss: 0.0003
    Epoch 44: Loss: 0.0003
    Epoch 45: Loss: 0.0002
    Epoch 46: Loss: 0.0002
    Epoch 47: Loss: 0.0002
    Epoch 48: Loss: 0.0002
    Epoch 49: Loss: 0.0002
    Epoch 50: Loss: 0.0002

**Minimal example:**

    import numpy as np
    from tinygrad.tensor import Tensor
    from tinygrad.nn import Conv2d, Linear, optim
    from tinygrad import Device

    def generate_training_data(batch_size, height=257, width=86, pos_fraction=0.3):
        X = np.random.randn(batch_size, 1, height, width).astype(np.float32)

        for i in range(batch_size):
            data = X[i, 0, :, :]
            data = (data - data.mean()) / (data.std() + 1e-8)
            X[i, 0, :, :] = data

        num_positives = int(batch_size * pos_fraction)
        y = np.zeros(batch_size, dtype=np.float32)
        y[:num_positives] = 1
        np.random.shuffle(y)

        X = Tensor(X)
        y = Tensor(y)
        return X, y

    print(f"Current device: {Device.DEFAULT}")
    X, y = generate_training_data(batch_size=10)

    Tensor.training = True

    # Define layers
    conv1 = Conv2d(1, 8, (3, 3))
    conv2 = Conv2d(8, 16, (3, 3))
    fc1 = Linear(19840, 64)
    fc2 = Linear(64, 1)
    params = (
        [conv1.weight, conv1.bias] +
        [conv2.weight, conv2.bias] +
        [fc1.weight, fc1.bias] +
        [fc2.weight, fc2.bias]
    )
    optimizer = optim.Adam(params, lr=0.001)

    def forward(x):
        x = conv1(x).relu().max_pool2d((2, 2))
        x = conv2(x).relu().max_pool2d((2, 2))
        x = x.reshape((x.shape[0], -1))
        x = fc1(x).relu()
        x = fc2(x).sigmoid()
        return x

    batch_size = 16
    epsilon = 1e-7
    for epoch in range(50):
        for i in range(0, len(X), batch_size):
            batch_X = X[i:i + batch_size]
            batch_y = y[i:i + batch_size]
            pred = forward(batch_X)
            pred = pred.clip(epsilon, 1 - epsilon)
            loss = -((batch_y.reshape(-1, 1) * pred.log()) + ((1 - batch_y.reshape(-1, 1)) * (1 - pred).log())).mean()
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
        print(f"Epoch {epoch + 1}: Loss: {loss.numpy():.4f}")
